src/compiler/checker.ts

```diff
@@ -17926,7 +18140,7 @@ namespace ts {
             return reduceLeft(types, (s, t) => isTypeSubtypeOf(t, s) ? t : s)!;
         }
 
-        function isArrayType(type: Type): boolean {
+        function isArrayType(type: Type): type is TypeReference {

```

Comment:
This shouldn't be a type guard (we haven't done #15048)

Reply:
Or rather the return type could be `type is (TypeReference & (something else))}` to avoid incorrect logic in `else` branches give a `TypeReference | Foo`

Reply:
In principle I agree, but the type guard does save several type casts. I suppose we could introduce an `ArrayTypeReference` and use that instead.